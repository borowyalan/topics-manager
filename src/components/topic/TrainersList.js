import React from 'react';
import Text from '../Text';
import styled from 'styled-components';


const TrainersList = props => (
    <StyledTrainersList>
        <Text type="secondary" margin="0 0 0.5em 0">Trenerzy</Text>
        <ul>
            <li>
                <Text>Piotr Zientara</Text>
            </li>
            <li>
                <Text>Ula Ka≈∫mierczyk</Text>
            </li>
            <li>
                <Text>Kamil Gajowy</Text>
            </li>
        </ul>
    </StyledTrainersList>
);

const StyledTrainersList = styled.figure`
    flex: 1;
    padding: 0 2em;
`;

export default TrainersList;
