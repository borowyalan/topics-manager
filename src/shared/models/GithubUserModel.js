import { UserModel } from './UserModel';
export class GithubUserModel extends UserModel {
    static fromOAuth0(idTokenPayload) {
        return Object.assign(new GithubUserModel(), idTokenPayload);
    }
}
