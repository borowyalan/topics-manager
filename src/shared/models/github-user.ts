import { User } from './user';

export interface GithubUser extends User {
    idTokenPayload: string;
}
