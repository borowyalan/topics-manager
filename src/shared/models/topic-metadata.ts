export interface TopicMetadata {
    title: string;
    description: string;
}
