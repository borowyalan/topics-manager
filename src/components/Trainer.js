import React from 'react';
import PropTypes from 'prop-types';
import {UserModel} from '../shared/models/UserModel';

const Trainer = ({trainer, important}) => (
    <div style={important ? {backgroundColor: 'yellow'} : {}}>
        <p>{trainer.name}</p>
    </div>
);

Trainer.propTypes = {
    trainer: PropTypes.instanceOf(UserModel).isRequired,
    important: PropTypes.bool
};

export default Trainer;
