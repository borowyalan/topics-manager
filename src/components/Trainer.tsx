import React, { SFC } from 'react';
import styled from 'styled-components';
import { User } from '../shared/models/user';
import Text, { TextType } from './Text';

export interface TrainerProps {
    trainer: User,
    important?: boolean,
}

const Trainer: SFC<TrainerProps> = ({ trainer, important = false }) => (
    <StyledTrainer>
        <Text
            type={TextType.Basic}
            style={{ display: 'inline', textDecoration: (important ? 'underline' : 'none') }}>
            {trainer.name}
        </Text>
    </StyledTrainer>
);

const StyledTrainer = styled.li`
    margin: 15px 0;
`;

export default Trainer;
