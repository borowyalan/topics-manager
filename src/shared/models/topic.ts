import { find } from 'lodash';
import { TopicMetadata } from './topic-metadata';
import { User } from './user';

export interface Topic extends TopicMetadata {
    trainers: User[];
    members: User[];
    author: User;
    id: string;
}

export const isAttending = (topic: Topic, user: User) => {
    return Boolean(user && find(topic.members, { email: user.email }));
};

export const isTrainer = (topic: Topic, user: User) => {
    return Boolean(
        user &&
            (topic.author.email === user.email ||
                find(topic.trainers, { email: user.email }))
    );
};
