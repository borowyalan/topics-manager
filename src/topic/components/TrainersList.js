import React from 'react';
import Text from '../../components/Text';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { UserModel } from '../../shared/models/user';
import Trainer from '../../components/Trainer';
import Topic from '../../shared/models/topic';
import BecomeTrainerContainer from '../BecomeTrainerContainer';
import Colors from '../../styles/colors';

const TrainersList = ({ trainers, author, topic }) => (
    <StyledTrainersContainer>
        <Text type="primary" margin="0 0 0.5em 0">
            Trenerzy
        </Text>
        <StyledTrainersList>
            <Trainer trainer={author} important />
            {trainers.map((trainer, index) => (
                <Trainer key={index} trainer={trainer} />
            ))}
        </StyledTrainersList>
        <BecomeTrainerContainer topic={topic} />
    </StyledTrainersContainer>
);

TrainersList.propTypes = {
    trainers: PropTypes.arrayOf(PropTypes.instanceOf(UserModel)).isRequired,
    topic: PropTypes.instanceOf(TopicModel).isRequired,
    author: PropTypes.instanceOf(UserModel).isRequired,
};

TrainersList.defaultProps = {
    trainers: [],
    author: UserModel.fromBackend({}),
};

const StyledTrainersContainer = styled.figure`
    flex: 1;
    padding: 0 2em;
`;

const StyledTrainersList = styled.ul`
    li::before {
        margin-right: 15px;
        content: 'â€¢';
        color: ${Colors.yellow};
    }
`;

export default TrainersList;
