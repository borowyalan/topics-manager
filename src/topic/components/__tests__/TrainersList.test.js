import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import 'jest-styled-components';
import TrainersList from '../TrainersList';
import Trainer from '../../../components/Trainer';
import { UserModel } from '../../../shared/models/user';

configure({ adapter: new Adapter() });

describe('<TrainersList>', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = shallow(<TrainersList />);
    });

    it('should render', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should have no additional trainers by default (only author)', () => {
        expect(wrapper.find(Trainer).length).toEqual(1);
    });

    describe('when trainers are provided', () => {
        const trainers = [UserModel.fromBackend({}), UserModel.fromBackend({})];
        beforeEach(() => {
            wrapper = shallow(<TrainersList trainers={trainers} />);
        });

        it('should match snapshot', () => {
            expect(wrapper).toMatchSnapshot();
        });

        it('should have exactly so much elements as passed to <TrainersList> + one (author)', () => {
            expect(wrapper.find(Trainer).length).toBe(trainers.length + 1);
        });
    });
});
