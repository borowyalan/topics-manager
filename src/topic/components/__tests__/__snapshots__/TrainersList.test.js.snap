// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrainersList> should render 1`] = `
<TrainersList__StyledTrainersContainer>
  <Text
    margin="0 0 0.5em 0"
    type="primary"
  >
    Trenerzy
  </Text>
  <TrainersList__StyledTrainersList>
    <Trainer
      important={true}
      trainer={
        UserModel {
          "email": undefined,
          "name": undefined,
          "nickname": undefined,
          "picture": undefined,
        }
      }
    />
  </TrainersList__StyledTrainersList>
  <Connect(BecomeTrainerContainer) />
</TrainersList__StyledTrainersContainer>
`;

exports[`<TrainersList> when trainers are provided should match snapshot 1`] = `
<TrainersList__StyledTrainersContainer>
  <Text
    margin="0 0 0.5em 0"
    type="primary"
  >
    Trenerzy
  </Text>
  <TrainersList__StyledTrainersList>
    <Trainer
      important={true}
      trainer={
        UserModel {
          "email": undefined,
          "name": undefined,
          "nickname": undefined,
          "picture": undefined,
        }
      }
    />
    <Trainer
      key="0"
      trainer={
        UserModel {
          "email": undefined,
          "name": undefined,
          "nickname": undefined,
          "picture": undefined,
        }
      }
    />
    <Trainer
      key="1"
      trainer={
        UserModel {
          "email": undefined,
          "name": undefined,
          "nickname": undefined,
          "picture": undefined,
        }
      }
    />
  </TrainersList__StyledTrainersList>
  <Connect(BecomeTrainerContainer) />
</TrainersList__StyledTrainersContainer>
`;
